src/pathfinding/grid_pathfinder.jl

```diff
@@ -0,0 +1,362 @@
+between two points in a `D` dimensional [`GridSpace{D}`](@ref).
+"""
+abstract type CostMetric{D} end
+
+struct DirectDistance{D} <: CostMetric{D}
+    direction_costs::Vector{Int}
+end
+
+Base.show(io::IO, metric::DirectDistance) = print(io, "DirectDistance")
+
+"""
+    DirectDistance
+Distance is approximated as the shortest path between the two points, provided the
+`walkable` property of [`AStar`](@ref) allows.
+Optionally provide a `Vector{Int}` that represents the cost of going from a tile to the
+neighboring tile on the `i` dimensional diagonal (default is `10√i`).
+
+If `moore_neighbors=false` in [`AStar`](@ref), only Von Neumann neighbors will be tested.
+Cost defaults to the first value of the provided vector.
+"""
+DirectDistance{D}() where {D} = DirectDistance{D}([floor(Int, 10.0 * √x) for x in 1:D])
+
+"""
+    Chebyshev
+Distance between two tiles is approximated as the Chebyshev distance (maximum of absolute
+difference in coordinates) between them.
+"""
+struct Chebyshev{D} <: CostMetric{D} end
+
+Base.show(io::IO, metric::Chebyshev) = print(io, "Chebyshev")
+
+struct HeightMap{D} <: CostMetric{D}
+    base_metric::CostMetric{D}
+    hmap::Array{Int,D}
+end
+
+Base.show(io::IO, metric::HeightMap) =
+    print(io, "HeightMap with base: $(metric.base_metric)")
+
+"""
+    HeightMap(hmap::Array{Int,D})
+    HeightMap(hmap::Array{Int,D}, ::Type{<:CostMetric})
+Distance between two positions is the sum of the shortest distance between them and the
+absolute difference in height. A heightmap of the same size as the corresponding
+[`GridSpace{D}`](@ref) is required. Distance is calculated using [`DirectDistance`](@ref)
+by default.
+"""
+HeightMap(hmap::Array{Int,D}) where {D} = HeightMap{D}(DirectDistance{D}(), hmap)
+
+HeightMap(hmap::Array{Int,D}, ::Type{M}) where {D,M<:CostMetric} =
+    HeightMap{D}(M{D}(), hmap)
+
+struct AStar{D,P,M} <: AbstractPathfinder
+    agent_paths::Dict{Int,Path{D}}
+    grid_dims::Dims{D}
+    neighborhood::Vector{CartesianIndex{D}}
+    admissibility::Float64
+    walkable::Array{Bool,D}
+    cost_metric::CostMetric{D}
+
+    function AStar{D,P,M}(
+        agent_paths::Dict,
+        grid_dims::Dims{D},
+        neighborhood::Vector{CartesianIndex{D}},
+        admissibility::Float64,
+        walkable::Array{Bool,D},
+        cost_metric::CostMetric{D},
+    ) where {D,P,M}
+        @assert admissibility >= 0 "Invalid value for admissibility: $admissibility ≱ 0"
+        @assert typeof(cost_metric) != HeightMap{D} || size(cost_metric.hmap) == grid_dims "Heightmap dimensions must be same as provided space"
+        new(agent_paths, grid_dims, neighborhood, admissibility, walkable, cost_metric)
+    end
+end
+
+"""
+    AStar(space::GridSpace; kwargs...)
+Provides pathfinding capabilities and stores agent paths. Dimensionality and periodicity
+properties are taken from `space`.
+
+`moore_neighbors = true` specifies if movement can be to Moore neighbors of a tile,
+or only Von Neumann neighbors.
```

Comment:
I am not a fan of this. Moore and Von Neumann are just not useful for someone who doesn't know the guys. Instead of using arbitrary names we should use names that convey what is happening. For example we could simply have the keyword be `movement_dirs` and the input a number `4` or `8` which then internally is converted to a Boolean. I'm not saying that my current suggestion is good, but I'm saying we should re-think this "Moore" stuff.

Reply:
That implies a 2D grid, but A* is N-dimensional. Need to be careful about that here too.

Reply:
orthogonal covers the "von Neumann" case in any dimnension I guess.